src/lib/operators/join_sort_merge.cpp

```diff
@@ -893,6 +1293,79 @@ class JoinSortMerge::JoinSortMergeImpl : public AbstractReadOnlyOperatorImpl {
         _output_pos_lists_right.push_back(std::move(null_output_right));
       }
     }
+
+    if (_mode == JoinMode::AntiNullAsTrue) {
+      // If the join mode is anti null as true with secondary predicates, we need to check for every left null row that:
+      // 1. There is no match with one of the right null rows where all predicates are satisfied.
+      // 2. There is no match with one of the left rows (not null) where all predicates are satisfied.
+      // We need to check the two cases because if  NULL AND X AND Y ... is for every combination False, we need
+      // to emit the row.
+      if (!_secondary_join_predicates.empty()) {
+        auto null_output_left = RowIDPosList();
+        const auto null_rows_left_size = _null_rows_left.size();
+        std::vector<bool> left_null_rows_matches(null_rows_left_size, false);
+        null_output_left.reserve(null_rows_left_size);
+
+        auto jobs = std::vector<std::shared_ptr<AbstractTask>>{};
+        const auto right_table_row_count = _right_input_table->row_count();
+
+        // We do not want to span to many threads. The upper limit is the number of CPU cores.
+        const auto num_cpus = Hyrise::get().topology.num_cpus();
+        size_t bunch_size = 1;
```

Comment:
If the join mode is anti null as true with secondary predicates, we need to check for every left null row that:
1. There is no match with one of the right null rows where all predicates are satisfied
2. There is no match with one of the left rows (not null) where all predicates are satisfied

We could do this by iterating over every single left row with a null value. My idea was that if we have a lot of rows to iterate over we could do this in parallel. For that, I divide the rows into bunches and these bunches are handled in separate threads. The size of a bunch is calculated with the number of CPUs.

This approach has some drawbacks:
1. If we have a lot of rows and for example 20 bunches and only the first row is a match, 19 bunches would do unneeded work.
2. I am not sure if there is even a case where we have so many null values where we would benefit from the parallel execution here.

Do you think we should keep it simple and just iterate in one thread over all null rows?

Reply:
Yes, please remove all threading code.

Actually, we don't allow operators to spawn threads on their own. That's what we have tasks for. But I don't see a good reason to have tasks here for this edge case. Let's tackle that in case it will ever become a bottleneck.